docs/source/concepts/ledger-model/ledger-exceptions.rst

```diff
@@ -0,0 +1,178 @@
+rollback node.
+
+So in the example of a consuming exercise in a rollback node, followed by a
+consuming exercise on the same contract outside of the rollback node, neither
+exercise comes "after" the other. They are part of separate continuities,
+one of which was rolled back, and so they don't break the ledger integrity.
+
+Privacy
++++++++
+
+Rollback nodes also have an interesting effect on the notion of privacy in
+the ledger model. When projecting a transaction for a party `p`, it's
+necessary to preserve some of the rollback structure of the transaction,
+even if `p` does not have the right to observe every action under it. This
+is because we need for `p` to be able to verify that an exercise under a
+rollback (to which they are is party) is conformant, but we also need `p`
+to know that the exercise was rolled back.
+
+As such, the notion of projection for rollback nodes is defined as follows:
+When projecting a rollback node for a party `p`, we project the subtransaction
+contained in the rollback node, and then we wrap the projected subtransaction
+in a new rollback node.
+
+Subsequently, we see whether we can "normalise" the projected rollback node to
+prevent leaking information about the structure of the original transaction.
+In particular, we apply the following steps:
+
+- If the projected rollback node starts with another rollback node, for instance:
+
+  .. code-block:: none
+
+    'Rollback' [ 'Rollback' tx , node1, ..., nodeN ]
+
+  Then we re-associate the rollback nodes, bringing the inner rollback node out:
+
+  .. code-block:: none
+
+    'Rollback' tx, 'Rollback' [ node1, ..., nodeN ]
+
+  We repeat this step until the projected rollback does not start with another
+  rollback node.
+
+- If the projected rollback node ends with another rollback node, for instance:
+
+  .. code-block:: none
+
+    'Rollback' [ node1, ..., nodeN, 'Rollback' [ node1', ..., nodeM' ] ]
+
+  Then we remove the flatten the inner rollback node into its parent:
+
+  .. code-block:: none
+
+    'Rollback' [ node1, ..., nodeN, node1', ..., nodeM' ]
+
+- If the projected rollback node is empty, we drop it.
+
+Note that all of these transformations preserve the "after" relation among
+actions in the transaction tree. They only affect the structure of the
+transaction by reducing the amount of variation that may occur in a
+projected rollback node. This way, information cannot be leaked through
+the arrangement of rollback nodes.
+
+The privacy section of the ledger model makes a point of saying that a
+ledger model should be **subaction-closed** to support projections. But
+this requirement is not necessarily true once we introduce rollbacks.
+Rollback nodes may contain actions that are not valid as standalone actions,
+since they may have been interrupted prematurely by an exception.
+
+Instead, we require that the ledger model be **projection-closed**, i.e.
+closed under projections for any party 'p'. This is a weaker requirement
```

Comment:
There's actually something more for Canton, which I also can't quite properly formulate yet :roll_eyes: I'll try to explain here. But I think it may be valuable if we can extract the essence, so worth spending some brain cycles on this.

For distributed checks we will have Alice checking what she thinks that Bob should see, and checking that Bob has actually seen it. Alice's view is based on `proj(tx, Alice)`, so she will be computing `proj(proj(tx, Alice), Bob)`. So this itself has to be model-conformant, otherwise, Alice might think that Bob would reject this.

There's probably a bit more that we want there. E.g., if Bob is "above" Alice in the transaction, then we probably want `proj(tx, Alice) == proj(proj(tx, Bob), Alice)`. But not sure how to formulate "above"...

Reply:
> For distributed checks we will have Alice checking what she thinks that Bob should see, and checking that Bob has actually seen it. Alice's view is based on proj(tx, Alice), so she will be computing proj(proj(tx, Alice), Bob). So this itself has to be model-conformant, otherwise, Alice might think that Bob would reject this.

This part should be implied by "projection-closed" I think. If it's closed under one projection, it's also closed under two projections!

Reply:
You're right, and I think the "privacy requirements" now mostly capture what I wanted to say (minus the Canton-distributed-check requirement, but we don't need to solve that for this PR)